### PR TITLE
Add -TemplateObject parameter to deployment cmdlets and re-record tests

### PR DESCRIPTION
**Pull request by cormacpayne**
## Description

Fix for https://github.com/Azure/azure-powershell/issues/2933

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/Resources/Resources/help/New-AzDeployment.md [hunk 1]
```diff
@@ -225,12 +253,27 @@ Accept pipeline input: True (ByPropertyName)
...
+Parameter Sets: ByTemplateObjectAndParameterObject, ByTemplateObjectAndParameterFile, ByTemplateObjectAndParameterUri, ByTemplateObjectWithNoParameters
+Aliases:
+
+Required: True
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
+
```
Comment (markcowl) [inline src/Resources/Resources/help/New-AzDeployment.md]: Please put a new example using the new parameter

File: src/Resources/Resources/help/New-AzResourceGroupDeployment.md [hunk 2]
```diff
@@ -302,6 +334,21 @@ Accept pipeline input: True (ByPropertyName)
...
+Type: System.Collections.Hashtable
+Parameter Sets: ByTemplateObjectAndParameterObject, ByTemplateObjectAndParameterFile, ByTemplateObjectAndParameterUri, ByTemplateObjectWithNoParameters
+Aliases:
+
+Required: True
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
```
Comment (markcowl) [inline src/Resources/Resources/help/New-AzResourceGroupDeployment.md]: add example as above

File: src/Resources/Resources/help/Test-AzDeployment.md [hunk 3]
```diff
@@ -150,12 +174,27 @@ Accept pipeline input: True (ByPropertyName)
...
+Type: System.Collections.Hashtable
+Parameter Sets: ByTemplateObjectAndParameterObject, ByTemplateObjectAndParameterFile, ByTemplateObjectAndParameterUri, ByTemplateObjectWithNoParameters
+Aliases:
+
+Required: True
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
```
Comment (markcowl) [inline src/Resources/Resources/help/Test-AzDeployment.md]: add example, as above

File: src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs [hunk 4]
```diff
@@ -433,7 +433,14 @@ private Deployment CreateBasicDeployment(PSDeploymentCmdletParameters parameters
...
             else
             {
-                deployment.Properties.Template = JObject.Parse(FileUtilities.DataStore.ReadFileAsText(parameters.TemplateFile));
+                if (!string.IsNullOrEmpty(parameters.TemplateFile))
+                {
+                    deployment.Properties.Template = JObject.Parse(FileUtilities.DataStore.ReadFileAsText(parameters.TemplateFile));
+                }
+                else
+                {
+                    deployment.Properties.Template = JObject.Parse(JsonConvert.SerializeObject(parameters.TemplateObject));
```
Comment (markcowl) [inline src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs]: Do we know that this is filled in at this point?  Do our parameter sets require this to be non-empty if we reach this point?

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 [hunk 5]
```diff
@@ -81,7 +111,7 @@ function Test-CrossResourceGroupDeploymentFromTemplateFile
 	$rgname = Get-ResourceGroupName
 	$rgname2 = Get-ResourceGroupName
 	$rname = Get-ResourceName
-	$rglocation = "CentralUSEUAP"
+	$rglocation = "West US 2"
```
Comment (markcowl) [inline src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1]: Can we fix this?  If not, can we file an issue to fix this?

File: src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs (hunk 4 above)
Comment (cormacpayne) [reply src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs]: @markcowl yes, `-TemplateFile`, `-TemplateUri` and `-TemplateObject` are all mutually exclusive parameters, and `parameters.TemplateFile` only has a value if the user provides a value to the first two parameters, and `parameters.TemplateObject` only has a value if the user provided a value to `-TemplateObject`, so neither property should be `null` when the user jumps into the respective block.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1 (hunk 5 above)
Comment (cormacpayne) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentTests.ps1]: @cormacpayne added comment about fixing these hard-coded locations across ARM tests in the issue that was filed for the other Resources PR: https://github.com/Azure/azure-powershell/issues/8678

File: src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs (hunk 4 above)
Comment (cormacpayne) [reply src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs]: @markcowl just added the `[ValidateNotNull]` attribute to the new `-TemplateObject` parameter

**Pull request merged**